wmi.go

```diff
@@ -109,9 +118,103 @@ type Client struct {
 	SWbemServicesClient *SWbemServices
 }
 
-// DefaultClient is the default Client and is used by Query, QueryNamespace
+// DefaultClient is the default Client and is used by Query, QueryNamespace, and CallMethod.
 var DefaultClient = &Client{}
 
+// coinitService coinitializes WMI service. If no error is returned, a cleanup function
+// is returned which must be executed (usually deferred) to clean up allocated resources.
+func (c *Client) coinitService(connectServerArgs ...interface{}) (*ole.IDispatch, func(), error) {
+	var unknown *ole.IUnknown
+	var wmi *ole.IDispatch
+	var serviceRaw *ole.VARIANT
+
+	// be sure teardown happens in the reverse
+	// order from that which they were created
+	deferFn := func() {
+		if serviceRaw != nil {
+			serviceRaw.Clear()
+		}
+		if wmi != nil {
+			wmi.Release()
+		}
+		if unknown != nil {
+			unknown.Release()
+		}
+		ole.CoUninitialize()
+	}
+
+	// if we error'ed here, clean up immediately
+	var err error
+	defer func() {
+		if err != nil {
+			deferFn()
+		}
+	}()
+
+	err = ole.CoInitializeEx(0, ole.COINIT_MULTITHREADED)
```

Comment:
Exactly, the coinitService return, the *Client, should be feed back into a global variable. The first time any method of this class is used, it will be initialized, but afterwards it can be reused until explicitly closed.

Reply:
Okay. Hmm. Given the complexity this is going to add, is there anything wrong or technically incorrect with initializing at each call and un-initializing after it? Otherwise, the cleanest way to do this will probably be to change the API of this package so that a client is "opened" (`wmi.OpenClient()` maybe, which returns a new Client ready to use), and when closed, then we close the client. I guess my question is, how hard do we need to enforce that only a single initialization is active at any given time?

If it's just for performance reasons, maybe we can be a little looser with the API.

If it's pretty important that initialization happens exactly once, and while active does not happen again, I will need to make a few more drastic changes to the API to guarantee that, and to ensure concurrent safety.

Reply:
@mholt that's definitely out of scope for this pr. I like leaving the initialization semantics as is.

Reply:
Its not that important, but re-opening it every time just pains a bit. If you prefer to keep it like this, that's fine and could be improved seperately.